### PR TITLE
Add webhook handler for auto-mounting projected `ServiceAccount` tokens into `Pod`s

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area security
/kind enhancement

**What this PR does / why we need it**:
This PR adds a webhook handler for auto-mounting projected `ServiceAccount` tokens into `Pod`s in case their referenced `ServiceAccount` sets `.automountServiceAccountToken=false`.

**Special notes for your reviewer**:
- ~✅ Based on #4817 which should be merged first~
- ~Only the last four commits are relevant (the others belong to #4817)~

**Release note**:

```feature operator
`gardener-resource-manager` can now auto-mount projected `ServiceAccount` tokens into `Pod`s in case their referenced `ServiceAccount` sets `.automountServiceAccountToken=false`. Please consult [this document](https://github.com/gardener/gardener/blob/master/docs/concepts/resource-manager.md#auto-mounting-projected-serviceaccount-tokens) for more information.
```

### CODE REVIEW

**Comment by rfranzke** [comment]
This PR is now
/ready
for review.
/squash

File: docs/concepts/resource-manager.md [hunk 1]
```diff
@@ -434,3 +434,43 @@ serviceaccount.resources.gardener.cloud/skip-deletion: "true"
...
+          name: kube-root-ca.crt
+      - downwardAPI:
+          items:
+          - fieldRef:
+              apiVersion: v1
+              fieldPath: metadata.namespace
+            path: namespace
+```
+
+> The `expirationSeconds` are defaulted to `12h` and can be overwritten with the `--projected-token-mount-expiration-seconds` flag.
```
Comment (timebertt) [inline docs/concepts/resource-manager.md]: Should we have an override for this via annotations on the Pod?

File: pkg/apis/resources/v1alpha1/types.go [hunk 2]
```diff
@@ -72,6 +72,10 @@ const (
 	DataKeyToken = "token"
 	// DataKeyKubeconfig is the data key whose value contains a kubeconfig with a service account token.
 	DataKeyKubeconfig = "kubeconfig"
+
+	// ProjectedTokenSkip is a constant for a label on a Pod which indicates that this Pod should not be considered for
```
Comment (timebertt) [inline pkg/apis/resources/v1alpha1/types.go:76]: Currently, it is used as an annotation, but the doc string says it's a label key.
However, I would actually go for the label, as we can then use it in an `objectSelector` in the `MutatingWebhookConfiguration`.

File: pkg/resourcemanager/webhook/projectedtokenmount/add.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+// AddToManager adds the webhook to a Manager using the default config.
+func AddToManager(mgr manager.Manager) error {
+	return AddToManagerWithOptions(mgr, defaultWebhookConfig)
+}
+
+// AddFlags adds the needed command line flags to the given FlagSet.
+func (o *WebhookOptions) AddFlags(fs *pflag.FlagSet) {
+	fs.Int64Var(&o.expirationSeconds, "projected-token-mount-expiration-seconds", 60*60*12, "expiration seconds for mounted projected service account tokens")
```
Comment (timebertt) [inline pkg/resourcemanager/webhook/projectedtokenmount/add.go:58]: Alternatively, we could use `fs.DurationVar`. WDYT?

File: pkg/resourcemanager/webhook/projectedtokenmount/handler.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	}
+}
+
+func (h *handler) InjectDecoder(d *admission.Decoder) error {
+	h.decoder = d
+	return nil
+}
+
+func (h *handler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	pod := &corev1.Pod{}
```
Comment (timebertt) [inline pkg/resourcemanager/webhook/projectedtokenmount/handler.go:60]: We only want to mutate `CREATE pods`, right? 
Should we check `req.Operation` accordingly?

File: pkg/resourcemanager/webhook/projectedtokenmount/handler.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+	if metav1.HasAnnotation(pod.ObjectMeta, resourcesv1alpha1.ProjectedTokenSkip) {
+		return admission.Allowed("pod explicitly opts out of projected service account token mount")
+	}
+
+	if len(pod.Spec.ServiceAccountName) == 0 || pod.Spec.ServiceAccountName == "default" {
+		return admission.Allowed("service account not specified or defaulted")
+	}
+
+	serviceAccount := &corev1.ServiceAccount{}
+	if err := h.targetClient.Get(ctx, kutil.Key(pod.Namespace, pod.Spec.ServiceAccountName), serviceAccount); err != nil {
```
Comment (timebertt) [inline pkg/resourcemanager/webhook/projectedtokenmount/handler.go:74]: Use the target cache explicitly?

File: pkg/resourcemanager/webhook/projectedtokenmount/handler.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	if serviceAccount.AutomountServiceAccountToken == nil || *serviceAccount.AutomountServiceAccountToken {
+		return admission.Allowed("auto-mounting service account token is not disabled")
+	}
+
+	for _, volume := range pod.Spec.Volumes {
+		if strings.HasPrefix(volume.Name, serviceAccountVolumeNamePrefix) {
+			return admission.Allowed("pod already has service account defaultVolume mount")
```
Comment (timebertt) [inline pkg/resourcemanager/webhook/projectedtokenmount/handler.go]: Can you clearify what is meant by `defaultVolume mount`?

File: pkg/resourcemanager/webhook/projectedtokenmount/handler.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+						ServiceAccountToken: &corev1.ServiceAccountTokenProjection{
+							ExpirationSeconds: &expirationSeconds,
+							Path:              "token",
+						},
+					},
+					{
+						ConfigMap: &corev1.ConfigMapProjection{
+							LocalObjectReference: corev1.LocalObjectReference{
+								Name: "kube-root-ca.crt",
+								// TODO: Use constant after https://github.com/gardener/gardener/pull/4832 is merged.
```
Comment (timebertt) [inline pkg/resourcemanager/webhook/projectedtokenmount/handler.go]: That PR is merged 😉

File: docs/concepts/resource-manager.md [hunk 8]
```diff
@@ -434,3 +434,43 @@ serviceaccount.resources.gardener.cloud/skip-deletion: "true"
...
+
+The volume will be mounted into all containers specified in the `Pod` to the path `/var/run/secrets/kubernetes.io/serviceaccount`.
```
Comment (timebertt) [inline docs/concepts/resource-manager.md:476]: Can you add a short sentence, why we are doing this?
I.e. explain that we are mimicking the service account admission plugin but with projected service account token volumes. Maybe link https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection, etc.

File: cmd/gardener-resource-manager/app/app.go [hunk 9]
```diff
@@ -56,6 +57,7 @@ func NewResourceManagerCommand() *cobra.Command {
 	tokenInvalidatorControllerOpts := &tokeninvalidatorcontroller.ControllerOptions{}
 	tokenRequestorControllerOpts := &tokenrequestorcontroller.ControllerOptions{}
 	rootCAControllerOpts := &rootcacontroller.ControllerOptions{}
+	projectedTokenMountOpts := &projectedtokenmountwebhook.WebhookOptions{}
```
Comment (timebertt) [inline cmd/gardener-resource-manager/app/app.go]: include `webhook` to follow the pattern?

**Review by timebertt — COMMENTED** [review]
Love it!
I mainly have some minor comments.

File: pkg/resourcemanager/webhook/projectedtokenmount/add.go (hunk 3 above)
Comment (rfranzke) [reply pkg/resourcemanager/webhook/projectedtokenmount/add.go:58]: The type in the `PodSpec` is `int64`, that's why I used it here as well. I would keep it for now.

Comment (timebertt) [reply pkg/resourcemanager/webhook/projectedtokenmount/add.go:58]: Ok, fine

File: docs/concepts/resource-manager.md [hunk 10]
```diff
@@ -434,3 +434,43 @@ serviceaccount.resources.gardener.cloud/skip-deletion: "true"
...
 
 then the respective `ServiceAccount` will not be deleted when the `Secret` is deleted.
+
+## Webhooks
+
+### Auto-Mounting Projected `ServiceAccount` Tokens
+
+When this webhook is activated then it automatically injects projected `ServiceAccount` token volumes into `Pod`s and all its containers if all of the following preconditions are fulfilled:
+
+1. The `Pod` is NOT annotated with `projected-token-mount.resources.gardener.cloud/skip=true`.
```
Comment (BeckerMax) [inline docs/concepts/resource-manager.md]:
```diff
1. The `Pod` is NOT labled with `projected-token-mount.resources.gardener.cloud/skip=true`.
```

File: pkg/resourcemanager/webhook/projectedtokenmount/handler.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+	if len(pod.Spec.ServiceAccountName) == 0 || pod.Spec.ServiceAccountName == "default" {
+		return admission.Allowed("service account not specified or defaulted")
+	}
+
+	serviceAccount := &corev1.ServiceAccount{}
+	if err := h.targetClient.Get(ctx, kutil.Key(pod.Namespace, pod.Spec.ServiceAccountName), serviceAccount); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	if serviceAccount.AutomountServiceAccountToken == nil || *serviceAccount.AutomountServiceAccountToken {
```
Comment (BeckerMax) [inline pkg/resourcemanager/webhook/projectedtokenmount/handler.go]: I saw you can also specify AutomountServiceAccountToken on the [pod level](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#use-the-default-service-account-to-access-the-api-server) which overwrites the SA setting. Should we also return here if that is set to `true` ?

File: pkg/resourcemanager/webhook/projectedtokenmount/handler.go [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+	return corev1.Volume{
+		Name: volumeName(),
+		VolumeSource: corev1.VolumeSource{
+			Projected: &corev1.ProjectedVolumeSource{
+				DefaultMode: pointer.Int32(420),
+				Sources: []corev1.VolumeProjection{
+					{
+						ServiceAccountToken: &corev1.ServiceAccountTokenProjection{
+							ExpirationSeconds: &expirationSeconds,
+							Path:              "token",
```
Comment (BeckerMax) [inline pkg/resourcemanager/webhook/projectedtokenmount/handler.go:133]: We are not using `audiences` here but I guess it doesn't provide additional value, or WDYT?

File: pkg/resourcemanager/webhook/projectedtokenmount/handler.go [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+							},
+							Items: []corev1.KeyToPath{{
+								// TODO: Use constant after https://github.com/gardener/gardener/pull/4832 is merged.
+								Key:  "ca.crt",
+								Path: "ca.crt",
+							}},
+						},
+					},
+					{
+						DownwardAPI: &corev1.DownwardAPIProjection{
```
Comment (BeckerMax) [inline pkg/resourcemanager/webhook/projectedtokenmount/handler.go:148]: I understand this adds the namespace of the pod to the volume. But why is this needed again?

File: pkg/resourcemanager/webhook/projectedtokenmount/handler.go [hunk 14]
```diff
@@ -50,12 +53,16 @@ func (h *handler) InjectDecoder(d *admission.Decoder) error {
 }
 
 func (h *handler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if req.Operation != admissionv1.Create {
```
Comment (BeckerMax) [inline pkg/resourcemanager/webhook/projectedtokenmount/handler.go:56]: What if a pod gets updated with an additional container? Then our volume mount will not be added to it. Is this fine?

File: docs/concepts/resource-manager.md [hunk 15]
```diff
@@ -434,3 +434,44 @@ serviceaccount.resources.gardener.cloud/skip-deletion: "true"
...
+          name: kube-root-ca.crt
+      - downwardAPI:
+          items:
+          - fieldRef:
+              apiVersion: v1
+              fieldPath: metadata.namespace
+            path: namespace
+```
+
+> The `expirationSeconds` are defaulted to `12h` and can be overwritten with the `--projected-token-mount-expiration-seconds` flag, or with the `projected-token-mount.resources.gardener.cloud/expiration-seconds` annotation on a `Pod` resource.
```
Comment (BeckerMax) [inline docs/concepts/resource-manager.md:474]: In the other PR (tokenRequestor) we thought about setting an upper limit of 24 hours on validity. Is this reasonable here too?

**Review by BeckerMax — CHANGES_REQUESTED** [review]
Thanks! Mostly general questions from my side.

File: pkg/resourcemanager/webhook/projectedtokenmount/handler.go (hunk 12 above)
Comment (rfranzke) [reply pkg/resourcemanager/webhook/projectedtokenmount/handler.go:133]: Yes, I agree, I don't think it would provide additional value. Apart from that, since the webhook would also be enabled for the seeds, using audiences wouldn't be easily possible anyways (w/o additional complexity) as we don't know which audiences are valid for a seed cluster.

File: pkg/resourcemanager/webhook/projectedtokenmount/handler.go (hunk 13 above)
Comment (rfranzke) [reply pkg/resourcemanager/webhook/projectedtokenmount/handler.go:148]: Mainly to maintain backwards-compatibility for "normal" service account secrets, and to maintain compatibility with the upstream `ServiceAccount` admission plugin: https://github.com/kubernetes/kubernetes/blob/37efc5feec6dacb233056dc0406d44060fb4c1f7/plugin/pkg/admission/serviceaccount/admission.go#L447-L458

Pods might use the `/var/run/secrets/kubernetes.io/serviceaccount/namespace` file for arbitrary purposes, so we should keep it even when using projected `ServiceAccount` tokens.

File: pkg/resourcemanager/webhook/projectedtokenmount/handler.go (hunk 14 above)
Comment (rfranzke) [reply pkg/resourcemanager/webhook/projectedtokenmount/handler.go:56]: Such updates are not allowed, see https://kubernetes.io/docs/concepts/workloads/pods/#pod-update-and-replacement:

> Pod updates may not change fields other than `spec.containers[*].image`, `spec.initContainers[*].image`, `spec.activeDeadlineSeconds` or `spec.tolerations`. For `spec.tolerations`, you can only add new entries.

File: docs/concepts/resource-manager.md (hunk 15 above)
Comment (rfranzke) [reply docs/concepts/resource-manager.md:474]: We didn't set such upper limit for the expiration duration of tokens. What we did was to set an upper limit of `24h` for the renew duration only, i.e., tokens valid for 1 month would still be renewed after 24 hours (instead of when 80% of the validity is reached).

The kubelet is already doing the same (from https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection):

>  The kubelet proactively rotates the token if it is older than 80% of its total TTL, or if the token is older than 24 hours.

File: pkg/resourcemanager/webhook/projectedtokenmount/handler.go (hunk 11 above)
Comment (rfranzke) [reply pkg/resourcemanager/webhook/projectedtokenmount/handler.go]: OK, done

File: docs/concepts/resource-manager.md (hunk 10 above)
Comment (rfranzke) [reply docs/concepts/resource-manager.md]: Good catch, thanks

File: docs/concepts/resource-manager.md (hunk 15 above)
Comment (BeckerMax) [reply docs/concepts/resource-manager.md:474]: got it. Thanks for the clarification.

**Review by BeckerMax — DISMISSED** [review]
/lgtm

**Comment by BeckerMax** [comment]
@rfranzke seems like  /go/src/github.com/gardener/gardener/pkg/resourcemanager/webhook/projectedtokenmount/handler_test.go:247 is failing in verify. Can you check?

**Comment by rfranzke** [comment]
@BeckerMax done

**Review by BeckerMax — APPROVED** [review]
/lgtm
